src/ble.rs

```diff
@@ -0,0 +1,179 @@
+        let adapter = adapters.first().context("no bluetooth adapter found")?;
+
+        adapter
+            .start_scan(ScanFilter {
+                services: vec![BADGE_SERVICE_UUID],
+            })
+            .await
+            .context("bluetooth scan start")?;
+        time::sleep(scan_duration).await;
+
+        // Filter for badge devices
+        let mut led_badges = vec![];
+        for p in adapter
+            .peripherals()
+            .await
+            .context("enumerating bluetooth devices")?
+        {
+            if let Some(badge) = Self::from_peripheral(p).await {
+                led_badges.push(badge);
+            }
+        }
+
+        Ok(led_badges)
+    }
+
+    async fn from_peripheral(peripheral: Peripheral) -> Option<Self> {
+        // Check whether the BLE device has the service UUID we're looking for
+        // and also the correct name.
+        // The service uuid is also by devices that are not LED badges, so
+        // the name check is also necessary.
+        let props = peripheral.properties().await;
+        if props.is_err() {
+            return None;
+        }
+
+        if let Some(props) = props.unwrap() {
+            let local_name = props.local_name.as_ref()?;
+            if local_name != BADGE_BLE_DEVICE_NAME {
+                return None;
+            }
+
+            if props
+                .services
+                .iter()
+                .any(|uuid| *uuid == BADGE_SERVICE_UUID)
+            {
+                Some(Self { peripheral })
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Return the single supported device
+    ///
+    /// This function returns an error if no device could be found
+    /// or if multiple devices would match.
+    pub async fn single() -> Result<Self> {
+        let mut devices = Self::enumerate()
+            .await
+            .context("enumerating badges")?
+            .into_iter();
+        let device = devices.next().context("no device found")?;
+        anyhow::ensure!(devices.next().is_none(), "multiple devices found");
+        Ok(device)
+    }
+
+    /// Write a payload to the device.
+    ///
+    /// This function connects to the device, writes the payload and disconnects.
+    /// When the device went out of range between discovering it
+    /// and writing the payload, an error is returned.
+    /// # Panics
+    /// This functions panics if the BLE device does not have the expected badge characteristic.
+    pub async fn write(&self, payload: PayloadBuffer) -> Result<()> {
+        self.peripheral
+            .connect()
+            .await
+            .context("bluetooth device connect")?;
+
+        let result = self.write_connected(payload).await;
+        if result.is_ok() {
+            self.peripheral
+                .disconnect()
+                .await
+                .context("bluetooth device disconnect")?;
+        }
```

Comment:
We want to always disconnect the peripheral, but ignore any errors during disconnect if we are already in an error state:
```suggestion
        let disconnect_result = self.peripheral
            .disconnect()
            .await;
        if result.is_ok() {
            disconnect_result.context("bluetooth device disconnect")?;
        } else {
            // ignore result of disconnect (an error already occurred)
        }
```